core/corehttp/gateway_handler.go

```diff
@@ -90,6 +90,17 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 
 	urlPath := r.URL.Path
 
+	// IPNSHostnameOption might have constructed an IPNS path using the Host header.
+	// In this case, we need the original path for constructing redirects
+	// and links that match the requested URL.
+	// For example, http://example.net would become /ipns/example.net, and
+	// the redirects and links would end up as http://example.net/ipns/example.net
+	originalUrlPath := urlPath
+	hdr := r.Header["X-IPNS-Original-Path"]
+	if len(hdr) > 0 {
+		originalUrlPath = hdr[0]
```

Comment:
maybe want to set here: `ipnsHostRedirection = true` and use that for the conditions below?